scripts/us_census/pep/monthly_population_estimate/preprocess.py

```diff
@@ -0,0 +1,369 @@
+        df.insert(0, 'date_range', _year_range(df['Date']), True)
+        return df
+
+    def _transform_data(self, df: pd.DataFrame) -> None:
+        """
+        This method calls the required functions to transform
+        the dataframe and saves the final cleaned data in
+        CSV file format.
+
+        Arguments:
+            file (str) : Dataset File Path
+
+        Returns:
+            df (DataFrame) : DataFrame.
+        """
+
+        # Finding the Dir Path
+        file_dir = os.path.dirname(self.cleaned_csv_file_path)
+        if not os.path.exists(file_dir):
+            os.mkdir(file_dir)
+        df = self._transform_df(df)
+        keep_value = "first"
+        if self.df is None:
+            self.df = df
+        else:
+            self.df = self.df.append(df, ignore_index=True)
+
+        self.df.sort_values(by=['Date', 'date_range'],
+                            ascending=False,
+                            inplace=True)
+        self.df.drop_duplicates("Date", keep=keep_value, inplace=True)
+        self.df.drop(['date_range'], axis=1, inplace=True)
+        self.df.to_csv(self.cleaned_csv_file_path, index=False)
+
+    def _generate_mcf(self, df_cols: list) -> None:
+        """
+        This method generates MCF file w.r.t
+        dataframe headers and defined MCF template
+
+        Arguments:
+            df_cols (list) : List of DataFrame Columns
+
+        Returns:
+            None
+        """
+        mcf_template = ("Node: dcid:{}\n"
+                        "typeOf: dcs:StatisticalVariable\n"
+                        "populationType: dcs:Person{}{}{}\n"
+                        "statType: dcs:measuredValue\n"
+                        "measuredProperty: dcs:count\n")
+        mcf = ""
+        for col in df_cols:
+            residence = ""
+            status = ""
+            armedf = ""
+            if col.lower() in ["date", "location"]:
+                continue
+            if re.findall('Resident', col):
+                residence = "\nresidentStatus"
+                status = ": dcs:USResident"
+            if re.findall('ArmedForces', col):
+                residence = "\nresidentStatus"
+                if len(status) == 0:
+                    status = ": dcs:InUSArmedForcesOverseas"
+                else:
+                    status = ": dcs:USResident__InUSArmedForcesOverseas"
+            if re.findall('Resides', col):
+                if re.findall('Household', col):
+                    residence = "\nresidenceType"
+                    status = ": dcs:Household"
+            if re.findall('NonInstitutionalized', col):
+                residence = "\ninstitutionalization"
+                status = ": dcs:USC_NonInstitutionalized"
+            if re.findall('Civilian', col):
+                armedf = "\narmedForcesStatus: dcs:Civilian"
+            if re.findall('Count_Person_InUSArmedForcesOverseas', col):
+                armedf = "\narmedForcesStatus: dcs:InArmedForces"

```

Comment:
For readability, you could do prepare lists of strings and do join.  Also move constants to globals, and use named args.  That is:

```
# Global
_MCF_TEMPLATE = """Node: dcid:{dcid}
typeOf: dcs:StatisticalVariable
populationType: dcs:Person
statType: dcs:measuredValue
measuredProperty: dcs:count
{xtra_pvs}
"""

def ...()
  ...
  mcf_nodes = []
  for col in df_cols:
    pvs = []
    if ...
      pvs.append("residenceType: dcs:Household")
    ...
    node = _MCF_TEMPLATE.format(dcid=col, xtra_pvs='\n'.join(pvs))
    mcf_nodes.append(node)

  mcf = '\n'.join(mcf_nodes)
```

Does that make sense?

Reply:
Made the code changes as per above suggestions.